### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by gavinpatkinson**
I refactored index.js into a more functional style. I did this originally to find a bug where identical keys were returned from `fullWalletFromSeed()` for different ships. This should make this library easier to read and understand. `fullWalletFromSeed()` is especially easier to read. All functions have been refactored and a few new helper functions have been added. Structurally, the wallet is almost nearly the same but I made two changes:

1) The ownership node simply uses `childNodeFromSeed()` directly with `ship` and `revision` set to `null`. This means ownership node will have a meta field, a ship field and a revision field, even though they are unused.

2) Several functions' args have changed to a config object which is destructured internally.

I also added a few helper functions:
- `bufferFrom()`: wraps `Buffer.from()`
- `bufferConcat()`: wraps `Buffer.concat()`
- `reverse()`: wraps `array.reverse()`
- `defaultTo(any, default)`: sets a default value if any is `undefined`
- `get()`: basically lodash `get()` but applies default if object itself is `undefined`.
- `isUndefined()`
- `isNumber()`

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+
+
+const argon2u = (entropy, ticketSize) => argon2({
+  pass: entropy, // string or Uint8Array
+  salt: 'urbitkeygen',
+  type: 10, // argon2.ArgonType.Argon2u,
+  hashLen: ticketSize,
+  // distPath: 'node_modules/argon2-wasm/dist',
```
Comment (Fang-) [inline src/index.js]: Does this actually work if you leave this commented out? If it does, great, feel free to remove this line.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+})
+
+
+
+const childSeedFromSeed = async config =>  {
+  const { seed, type, revision, ship, password } = config
+
+  const salt = typeof ship === 'number'
+    ? `${salt}-${ship}`
+    : `${type}-${revision}`
```
Comment (Fang-) [inline src/index.js]: This... doesn't match the previous implementation, I don't think? On line 54, `salt` is still empty, correct? We want `type-revision` always, and then `-ship` if we have one. If I'm reading this correctly it just gives you either `-ship` or `type-revision`.

Comment (gavinpatkinson) [reply src/index.js]: You're right, this isn't correct.

File: src/index.js (hunk 1 above)
Comment (gavinpatkinson) [reply src/index.js]: Yes, it still works.

**Comment by gavinpatkinson** [comment]
The semicolons are only occasionally necessary, but it is very much reader preference. Ted once told me he prefers semis because it clarifies when an expression ends. On the other hand, I'm very used to reading JS w/o semis.

**Comment by cgyarvin** [comment]
We should definitely pick a common, well-known, modern JS convention... not the expert here though.  But I thought most JS people fear the implicit semicolon...

**Comment by gavinpatkinson** [comment]
@c-johnson probably knows more about js semicolons than I do.

**Comment by cgyarvin** [comment]
> dog in this holy war

I’m reminded of a time early in my career when I wanted to express my dissatisfaction with improving an overly mature codebase.  I found myself describing this futile process as “putting lipstick on a dead horse.”

**Comment by gavinpatkinson** [comment]
Ok, semis are in, I also commented all the functions.

**Comment by vvisigoth** [comment]
If only all holy wars were this civil. Merging.

**Comment by Fang-** [comment]
But he hadn't removed the `// distPath` yet! ):

That's fine though. I'm going to assume the answer to this question of mine was yes:

> does the output from this match the output from the old version, when asking for just a single ship?

If not, there's an issue in the hoon version of this, too...

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+  // Normalize revisions object
+  const _revisions = {
+    transfer: get(revisions, 'transfer', 0),
+    spawn: get(revisions, 'spawn', 0),
+    delegate: get(revisions, 'delegate', 0),
+    manage: get(revisions, 'manage', 0),
+    network: get(revisions, 'network', 0),
+  };
+
+  const ownershipNode = await childNodeFromSeed({
```
Comment (Fang-) [inline src/index.js]: @gavinpatkinson Actually, I realize now that this isn't correct, and doesn't match the original implementation. What you're doing here is deriving a seed from the `ownerSeed`, and then getting a wallet from that. What you really want to be doing (according to previous implementation and design) is to derive a wallet from the `ownerSeed` directly.

**Pull request merged**